packages/stylelint-plugin/src/rules/no-invalid-use-of-tokens.ts

```diff
@@ -0,0 +1,354 @@
+      })
+      return
+    }
+
+    const commitReplacement = () => {
+      changed = true
+      if (!options.fix) {
+        report({
+          message: deprecatedTokenUsageMessage(variable.name, replacement),
+          autofixAvailable: true,
+        })
+      } else {
+        // Here is where we actually commit the replacement
+
+        const replacementVariable = replaceTokenInVariable(
+          variable,
+          replacementKaizenToken
+        )
+        variableNode.value = stringifyVariable(replacementVariable)
+
+        newValue = postcssValueParser.stringify(parsedValue.nodes)
+      }
+    }
+
+    // Now decide whether an autofix can occur using a series of rules.
+    // If we decide that it can't happen, report the reason then return so that we move on to the next node.
+
+    const isCurrentTokenCssVariable = !!variable.kaizenToken?.cssVariable
+    const isReplacementTokenCssVariable = !!replacementKaizenToken.cssVariable
+
+    // If the current value is a CSS variable, OR the replacement is not a CSS variable, just commit the replacement because it shouldn't be a problematic change.
+    if (isCurrentTokenCssVariable || !isReplacementTokenCssVariable) {
+      commitReplacement()
+      return
+    }
+
+    // If the potential replacement variable is a CSS variable, and is used within a function that doesn't support CSS variables, report and bail
+    if (
+      parent &&
+      !functionsThatSupportCSSVariables.has(parent.value) &&
+      replacementKaizenToken.cssVariable
+    ) {
+      report({
+        message: replacementCssVariableUsedWithinUnsupportedFunction(
+          variable.name,
+          replacementKaizenToken.name,
+          parent.value
+        ),
+        autofixAvailable: false,
+      })
+
+      return
+    }
+
+    // If we're in an AtRule that doesn't support CSS variables, and the replacement is a CSS variable, we shouldn't make the replacement.
+    if (
+      postcssNode.type === "atrule" &&
+      replacementKaizenToken.cssVariable &&
+      atRulesThatDontSupportCSSVars.has(postcssNode.name)
+    ) {
+      report({
+        message: cantReplaceDeprecatedTokenInAtRuleMessage(
+          variable.name,
+          replacementKaizenToken.name
+        ),
+        autofixAvailable: false,
+      })
+      return
+    }
+
+    // We can only modify variable declarations (e.g. `$card-bg: $kz-var-color-wisteria-100;`) when its current value is a CSS variable, or the replacement isn't a CSS variable
+    // (because the only change which isn't safe is going from NonCSSVariable -> CSSVariable)
+    // The process will only reach here if a safe change is not available, due to the checks above.
+    // If there isn't, just report, and bail.
+    if (postcssNode.type === "decl" && isVariable(postcssNode)) {
+      report({
+        message: deprecatedTokenInVariableMessage(variable.name, replacement),
+        autofixAvailable: false,
+      })
+      return
+    }
```

Comment:
Moving this rule up so it comes directly after the check about if the current value _is_ a CSS Variable, or if the new values _is not_ a CSS variable, will make it easier to match what the comment is saying to what the control flow in the code is doing.